pkg/cmd/flags/filter.go

```diff
@@ -128,203 +98,271 @@ func PrepareFilter(filtersArr []string) (tracee.Filter, error) {
 		}
 	}
 
-	for _, filterFlag := range filtersArr {
-		filterName := filterFlag
-		operatorAndValues := ""
-		operatorIndex := strings.IndexAny(filterFlag, "=!<>")
-		if operatorIndex > 0 {
-			filterName = filterFlag[0:operatorIndex]
-			operatorAndValues = filterFlag[operatorIndex:]
-		}
-
-		if len(operatorAndValues) == 1 || operatorAndValues == "!=" || operatorAndValues == "<=" || operatorAndValues == ">=" {
-			return tracee.Filter{}, filters.InvalidExpression(filterFlag)
+	processFilterFlag := func(filterFlag string) (filterName, operatorAndValues string, operatorIdx, scopeIdx int, err error) {
+		scopeIdx = 1
+		operatorIdx = strings.IndexAny(filterFlag, "=!<>")
+		if operatorIdx == -1 {
+			return "", "", 0, 0, filters.InvalidExpression(filterFlag)
 		}
 
-		if strings.Contains(filterFlag, ".retval") {
-			err := filter.RetFilter.Parse(filterName, operatorAndValues, eventsNameToID)
-			if err != nil {
-				return tracee.Filter{}, err
+		dashIndex := strings.LastIndex(filterFlag[0:operatorIdx], "-")
+		if dashIndex != -1 {
+			if dashIndex+1 >= len(filterFlag) {
+				return "", "", 0, 0, filters.InvalidScope(filterFlag)
 			}
-			continue
-		}
-
-		if strings.Contains(filterFlag, ".context") {
-			err := filter.ContextFilter.Parse(filterName, operatorAndValues)
+			scopeIdx, err = strconv.Atoi(filterFlag[dashIndex+1 : operatorIdx])
 			if err != nil {
-				return tracee.Filter{}, err
+				return "", "", 0, 0, filters.InvalidScope(fmt.Sprintf("%s - %s", filterFlag, err))
 			}
-			continue
-		}
-
-		if strings.Contains(filterFlag, ".args") {
-			err := filter.ArgFilter.Parse(filterName, operatorAndValues, eventsNameToID)
-			if err != nil {
-				return tracee.Filter{}, err
+			if scopeIdx < 1 || scopeIdx > tracee.MaxFilterScopes {
+				return "", "", 0, 0, filters.InvalidScope(fmt.Sprintf("%s - scopes must be between 1 and %d", filterFlag, tracee.MaxFilterScopes))
 			}
-			continue
+		} else {
+			dashIndex = operatorIdx
 		}
 
-		// The filters which are more common (container, event, pid, set, uid) can be given using a prefix of them.
-		// Other filters should be given using their full name.
-		// To avoid collisions between filters that share the same prefix, put the filters which should have an exact match first!
-		if filterName == "comm" {
-			err := filter.CommFilter.Parse(operatorAndValues)
-			if err != nil {
-				return tracee.Filter{}, err
-			}
-			continue
+		filterName = filterFlag[0:dashIndex]
+		operatorAndValues = filterFlag[operatorIdx:]
+		scopeIdx--
```

Comment:
I see that no scope and scope -1 is "scope 0" actually. Why not to have "filter-0", "filter-1", "filter-2" and make "filter" or "filter-0" the base ones ?

Reply:
Probably because the end user would be confused about adding a "filter=X" then a "filter-1=X", right? (Since it would be the "second" filter scope). Alright.